stl/inc/ranges

```diff
@@ -2490,6 +2490,412 @@ namespace ranges {
+                    return _Tmp;
+                } else {
+                    ++*this;
+                }
+            }
+
+            _NODISCARD friend constexpr bool operator==(const _Outer_iter& _Left, const _Outer_iter& _Right) noexcept(
+                noexcept(_Left._Current == _Right._Current)) /* strengthened */ requires forward_range<_BaseTy> {
+                return _Left._Current == _Right._Current;
+            }
+            _NODISCARD friend constexpr bool operator==(const _Outer_iter& _Left, default_sentinel_t) noexcept(
+                noexcept(_Left._At_end())) /* strengthened */ {
+                return _Left._At_end();
+            }
+        };
+
+        template <bool _Const>
+        class _Inner_iter {
+        private:
+            using _BaseTy = _Maybe_const<_Const, _Vw>;
+
+            _Outer_iter<_Const> _It{};
+            bool _Incremented = false;
+
+            _NODISCARD constexpr bool _Equal(const _Inner_iter& _Right) const {
+                return _It._Get_current() == _Right._It._Get_current();
+            }
+
+            _NODISCARD constexpr iterator_t<_BaseTy>& _Get_current() noexcept {
+                return _It._Get_current();
+            }
+
+            _NODISCARD constexpr const iterator_t<_BaseTy>& _Get_current() const noexcept {
+                return _It._Get_current();
+            }
+
+            _NODISCARD constexpr bool _At_end() const {
+                auto _Pat_pos       = _RANGES begin(_It._Parent->_Pattern);
+                const auto _Pat_end = _RANGES end(_It._Parent->_Pattern);
+                auto _Last          = _RANGES end(_It._Parent->_Range);
+                if constexpr (_Tiny_range<_Pat>) {
+                    const auto& _Cur = _It._Get_current();
+                    if (_Cur == _Last) {
+                        return true;
+                    }
+
+                    if (_Pat_pos == _Pat_end) {
+                        return _Incremented;
+                    }
+                    return *_Cur == *_Pat_pos;
+                } else {
+                    auto _Cur = _It._Get_current();
+                    if (_Cur == _Last) {
+                        return true;
+                    }
+
+                    if (_Pat_pos == _Pat_end) {
+                        return _Incremented;
+                    }
+
+                    do {
+                        if (*_Cur != *_Pat_pos) {
+                            return false;
+                        }
+
+                        if (++_Pat_pos == _Pat_end) {
+                            return true;
+                        }
+                    } while (++_Cur != _Last);
+                    return false;
+                }
+            }
+
+            using _BaseCategory = typename iterator_traits<iterator_t<_BaseTy>>::iterator_category;
+
+        public:
+            using iterator_concept = typename _Outer_iter<_Const>::iterator_concept;
+            using iterator_category =

```

Comment:
Per WG21-P2259, this `iterator_category` is present only if _BaseTy models forward_range

Reply:
@timsong-cpp is it intentional that [range.split.inner]/1, after applying P2256R1, tells me what type `iterator_category` denotes in cases where it is not present? I'm not sure what it means for a name that doesn't exist to denote some type. Did LWG consider making the change:
```diff
-1. The
+1. When present, the
 typedef-name iterator_­category denotes:
```
for clarity?

Reply:
I don't remember anyone bringing this up. I think we can make that change editorially?

Reply:
> I don't remember anyone bringing this up. I think we can make that change editorially?

Assuming that we can convince the Editor that the meaning of a name that doesn't exist is not normative =)

Reply:
> I'll prepare an editorial PR. I figured I'd ask first just in case I was missing something.

This is now cplusplus/draft#4545.